src/Core/Services/ExecutionHelper.cs

```diff
@@ -0,0 +1,517 @@
+                            variables: variables));
+                }
+            }
+
+            // Overwrite the default values with the passed in arguments
+            IReadOnlyList<ArgumentNode> passedArguments = query.Arguments;
+
+            foreach (ArgumentNode argument in passedArguments)
+            {
+                string argumentName = argument.Name.Value;
+                IInputField argumentSchema = argumentSchemas[argumentName];
+
+                if (parameters.ContainsKey(argumentName))
+                {
+                    parameters[argumentName] =
+                        ExtractValueFromIValueNode(
+                            value: argument.Value,
+                            argumentSchema: argumentSchema,
+                            variables: variables);
+                }
+                else
+                {
+                    parameters.Add(
+                        argumentName,
+                        ExtractValueFromIValueNode(
+                            value: argument.Value,
+                            argumentSchema: argumentSchema,
+                            variables: variables));
+                }
+            }
+
+            return parameters;
+        }
+
+        /// <summary>
+        /// InnerMostType is innermost type of the passed Graph QL type.
+        /// This strips all modifiers, such as List and Non-Null.
+        /// So the following GraphQL types would all have the underlyingType Book:
+        /// - Book
+        /// - [Book]
+        /// - Book!
+        /// - [Book]!
+        /// - [Book!]!
+        /// </summary>
+        internal static IType InnerMostType(IType type)
+        {
+            if (type.ToString() == type.InnerType().ToString())
+            {
+                return type;
+            }
+
+            return InnerMostType(type.InnerType());
+        }
+
+        public static InputObjectType InputObjectTypeFromIInputField(IInputField field)
+        {
+            return (InputObjectType)(InnerMostType(field.Type));
+        }
+
+        private static IDictionary<string, object?> GetParametersFromContext(
+            IMiddlewareContext context)
+        {
+            return GetParametersFromSchemaAndQueryFields(
+                context.Selection.Field,
+                context.Selection.SyntaxNode,
+                context.Variables);
+        }
+
+        /// <summary>
+        /// Get metadata from context.
+        /// The metadata key is the root field name + _PURE_RESOLVER_CTX + :: + PathDepth.
+        /// </summary>
+        private static IMetadata GetMetadata(IPureResolverContext context)
+        {
+            if (context.Selection.ResponseName == QueryBuilder.PAGINATION_FIELD_NAME && context.Path.Parent is not null)
+            {
+                string paginationObjectParentName = GetMetadataKey(context.Path) + "::" + context.Path.Parent.Depth;
+                return (IMetadata)context.ContextData[paginationObjectParentName]!;
```

Comment:
Yes, however after review, I prefer not to use the null forgiving operator. 
updating to throw an exception if the exceptional situation of the metadata is null. 
```csharp
if (!context.ContextData.TryGetValue(paginationObjectParentName, out object? response) || response is null)
{
     throw new DataApiBuilderException(
        statusCode: HttpStatusCode.BadRequest,
        subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedError,
        message: $"Pagination metadata for {context.Path} not found in context.ContextData.");
}

return (IMetadata)response;
```

Reply:
Yes for SQL, no for Cosmos as cosmos doesn't use Imetadata. The challenge here is that both cosmos and sql query engines implement an interface that mandates parameter of Imetadata. I removed null forgiving operator in favor of a nullable IMetadata object. Can't throw an exception if `context.ContextData[paginationObjectParentName]` because that would break cosmos.